sdk/src/main/scala/com/horizen/account/state/ForgerStakeMsgProcessor.scala

```diff
@@ -366,37 +387,147 @@ case class ForgerStakeMsgProcessor(params: NetworkParams) extends FakeSmartContr
     stakeId
   }
 
+  private def getAllowedForgersIndexList(view: BaseAccountStateView): Array[Byte] = {
+
+    if (networkParams.allowedForgersList.isEmpty){
+      throw new IllegalStateException("Illegal call when list of forger is empty")
+    }
+
+    // get the forger list. Lazy init
+    val restrictForgerList = view.getAccountStorage(contractAddress, RestrictedForgerFlagsList)
+    if (restrictForgerList.sameElements(NULL_HEX_STRING_32)) {
+      // it is the first time we access this item, do the init now
+      val forgersIndexesArray = new Array[Byte](networkParams.allowedForgersList.size)
+      view.updateAccountStorageBytes(contractAddress, RestrictedForgerFlagsList, forgersIndexesArray)
+    }
+    view.getAccountStorageBytes(contractAddress, RestrictedForgerFlagsList)
+
+  }
+
+  def doOpenStakeForgerListCmd(msg: Message, view: BaseAccountStateView): Array[Byte] = {
+
+    if (!networkParams.restrictForgers) {
+      throw new ExecutionRevertedException("Illegal call when list of forger is not restricted")
+    }
+
+    if (networkParams.allowedForgersList.isEmpty){
+      throw new ExecutionRevertedException("Illegal call when list of forger is empty")
+    }
+
+    if (msg.getNonce == null) {
+      throw new ExecutionRevertedException("Call must include a nonce")
+    }
+
+    if (msg.getValue.signum() != 0) {
+      throw new ExecutionRevertedException("Call value must be zero")
+    }
+
+    val inputParams = getArgumentsFromData(msg.getData)
+    val cmdInput = OpenStakeForgerListCmdInputDecoder.decode(inputParams)
+    val forgerIndex: Int = cmdInput.forgerIndex
+    val signature: Signature25519 = cmdInput.signature
+
+    // get the forger list. Lazy init
+    val restrictForgerList = getAllowedForgersIndexList(view)
+
+    // check consistency of input.
+    if (networkParams.allowedForgersList.size < forgerIndex+1) {
+      throw new ExecutionRevertedException(
+        s"Invalid forgerIndex=$forgerIndex: allowedForgersList size=${networkParams.allowedForgersList.size}")
+    }
+
+    // check signature
+    val blockSignerProposition = networkParams.allowedForgersList(forgerIndex)._1
+
+    val msgToSign = getOpenStakeForgerListCmdMessageToSign(forgerIndex, msg.getFrom.address(), msg.getNonce.toByteArray)
+    if (!signature.isValid(blockSignerProposition, msgToSign)) {
+      throw new ExecutionRevertedException(s"Invalid signature, could not validate against blockSignerProposition=$blockSignerProposition")
+    }
+
+    // check that the forger list is not already open
+    if (isForgerListOpenUnchecked(restrictForgerList)) {
+      throw new ExecutionRevertedException("Forger list already open")
+    }
+
+    // check that the index has not been already processed
+    if (restrictForgerList(forgerIndex) == 1) {
+      throw new ExecutionRevertedException("Forger index already processed")
+    }
```

Comment:
I left as it is because moving before signature validation would consume some gas in case the signature verification fails. In other words I prefer gas saving over speed, but we may discuss it.

Reply:
Ah ok didn't know that, thanks